File: go/day1/main.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"golang.org/x/exp/slices"
)

func toInt(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

func parseInput(input string) (res [2][]int) {
	for _, s := range strings.Split(input, "\n") {
		a, b, _ := strings.Cut(s, "   ")
		res[0] = append(res[0], toInt(a))
		res[1] = append(res[1], toInt(b))
	}
	return res
}

func SumDistance(input string) (sum int) {
	res := parseInput(input)

	slices.Sort(res[0])
	slices.Sort(res[1])

	for i := range res[0] {
		sum += int(math.Abs(float64(res[0][i] - res[1][i])))
	}

	return sum
}

func main() {
	fmt.Println("Part 1:", SumDistance(input))
}

File: ts/day1/main.ts
import { input } from "./input.ts";

const parseInput = (input: string): [number[], number[]] => 
  input.split("\n")
    .reduce<[number[], number[]]>((acc, line) => {
      const [l, r] = line.split(/\s+/)
      acc[0].push(parseInt(l, 10));
      acc[1].push(parseInt(r, 10));
      return acc;
    } , [[], []]);

export const sumDistance = (input: string): number => {
  const [left, right] = parseInput(input);

  const l = left.sort()
  const r = right.sort()

  return l.reduce((sum, a, i) => sum + Math.abs(a - r[i]), 0)
}

export const similarityScore = (input: string): number => {
  const [left, right] = parseInput(input);

  return left.reduce((score, a) => score + (a * right.filter(b => a === b).length), 0)
}

// Learn more at https://docs.deno.com/runtime/manual/examples/module_metadata#concepts
if (import.meta.main) {
  console.log("Part 1:", sumDistance(input))
  console.log("Part 2:", similarityScore(input))
}
